data-pipeline/src/covid_variant_pipeline/assign_clades.py

```diff
@@ -116,6 +137,43 @@ def assign_clades(as_of_date: str, reference_tree: AnyPath, root_sequence: AnyPa
     return assignment_file
 
 
+def merge_metadata(
+    as_of_date: str, run_datetime: datetime.datetime, metadata_file: AnyPath, assignment_file: AnyPath
+) -> pl.DataFrame:
+    """Merge sequence metadata with clade assignments."""
+
+    df_metadata = pl.read_csv(metadata_file, separator="\t")
+
+    # we're expecting one row per sequence id (aka Accession)
+    # TODO: how do we want to handle the case where the metadata file has
+    # duplicate Accession values?
+    assert df_metadata["Accession"].n_unique() == df_metadata.shape[0]
+
+    df_assignments = pl.read_csv(assignment_file, separator=";")
+    df_assignments = parse_sequence_assignments(df_assignments)
+
+    joined = df_metadata.join(df_assignments, left_on="Accession", right_on="seq", how="left")
+    joined = joined.with_columns(
+        sequence_retrieved_datetime=run_datetime,
+        reference_tree_date=datetime.datetime.strptime(as_of_date, "%Y-%m-%d").date(),
+    )
+    num_sequences = joined.shape[0]
+
+    missing_clade_assignments = joined.filter(pl.col("Accession").is_null())
```

Comment:
I have two questions on this part (that I'll ask in three pieces):

1. What do we know about the metadata file and the sequence file:
    1. do they always have the same number of entries, for the same set of sequences?
    2. might the metadata file have stuff that's not in the sequence file?
    3. might the sequence file have stuff that's not in the metadata file?
2. Question 1 was a preliminary for understanding why we're doing a left join of assignments into metadata.  That seems safe if we know that we're not in case 1.iii.  Otherwise, a right join or outer join might be safer?
3. Not having dug into how polars works, my interpretation of what's happening in line 162 is that we're keeping rows of the joined df where the Accession is missing.  Based on the left join operation, I would have expected that the filter operation we'd want would be for rows that didn't end up being assigned a nextstrain clade, i.e. something like `pl.col("clade_nextstrain").is_null()`

Reply:
good questions...will tackle in separate comments, starting with the first

For this iteration, I was working under the assumption that because the sequence data and the sequence metadata are part of the same NCBI "package" that they would contain the same number of entries.

Using [this method for determining the number of sequence in a fasta file,](https://bioinformaticsworkbook.org/Appendix/Unix/unix-basics-3grep.html#gsc.tab=0) and using the files I downloaded in a pipeline run timestamped `20240513T142649`, we can see this is true:

```python
import os
import polars as pl

t = '20240513T142649'
df_metadata = pl.read_csv(f'data/sequence/{t}/{t}-metadata.tsv', separator='\t')
fastafasta = f'data/sequence/{t}/ncbi_dataset/data/genomic.fna'

df_metadata.shape
# (6026, 19)

num_sequences = os.system(f'grep -c "^>" {fastafasta}')
# 6026
```

I love your question because it's never good to assume anything, really. Might the sequences and their metadata differ? They might, and we should check for that.

What do you think we should do if the numbers don't match?

1. Stop the pipeline with an error (i.e., don't continue with the clade assignment process)
2. Continue the pipeline with a warning
3. ¯\_(ツ)_/¯

My sense is that we'd care more about sequences that don't have metadata than the opposite?

FWIW, we're not currently parsing the sequence data directly (i.e., there's no nicely formatted dataframe that represents the fasta file), so we'd need to do that if want to report on specific sequences that don't have a match.

Reply:
For question 2, the prior assumption is indeed the reason for the left join.

Assuming that the metadata file contains all sequences that we care about, we want to maintain them in the final file, even if they don't have a clade assignment.

Reply:
Question 3 -- you are right, and that's a great catch! I'll make a commit to fix this....appreciate the close read.

Reply:
thanks!

question 1: i agree, it seems like we care more about sequences without metadata than metadata without sequences.  maybe if there were sequences without metadata we could continue with a warning?  that would let humans look into it to see how serious the issue is afterwards, with results to work with in case they're ok.

question 2: makes sense

question 3: great!

should we file a separate issue for 1/2?

Reply:
https://github.com/reichlab/variant-nowcast-hub/issues/14